Error Building GMP

`./cross_compile_ffmpeg.sh`, select Win32 at prompt, wait until it blows.

```
2025-12-26 13:16:11 URL:http://ftp.gnu.org/gnu/gmp/gmp-6.2.1.tar.xz [2027316/2027316] -> "gmp-6.2.1.tar.xz" [1]
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... ../source/gmp-6.2.1/install-sh -c -d
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether to enable maintainer-specific portions of Makefiles... no
checking ABI=64
checking compiler gcc -O2 -fexceptions... no, long long reliability test 1
checking ABI=x32
checking compiler gcc -O2 -fexceptions... no, long long reliability test 1
checking ABI=32
checking compiler gcc -O2 -fexceptions... no, long long reliability test 1
configure: error: could not find a working compiler, see config.log for details
error

See build.log for more details.
Failure building 32 bit gcc? Recommend nuke sandbox (rm -rf sandbox) and start over...
```

```
Test compile: __builtin_alloca availability
configure:6354: gcc -O2 -pedantic -fomit-frame-pointer -m64  conftest.c >&5
conftest.c: In function 'foo':
conftest.c:1:21: warning: ignoring return value of '__builtin_alloca' declared with attribute 'warn_unused_result' [-Wunused-result]
    1 | int k; int foo () { __builtin_alloca (k); }
      |                     ^~~~~~~~~~~~~~~~~~~~
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/15/../../../x86_64-linux-gnu/Scrt1.o: in function `_start':
(.text+0x1b): undefined reference to `main'
collect2: error: ld returned 1 exit status
configure:6357: $? = 1
failed program was:
int k; int foo () { __builtin_alloca (k); }
```

```
Test compile: long long reliability test 1
configure:6543: gcc -O2 -pedantic -fomit-frame-pointer -m64  conftest.c >&5
conftest.c: In function 'f':
conftest.c:12:48: error: too many arguments to function 'g'; expected 0, have 6
   12 | for(i=0;i<1;i++){if(e(got,got,9,d[i].n)==0)h();g(i,d[i].src,d[i].n,got,d[i].want,9);if(d[i].n)h();}}
      |                                                ^ ~
conftest.c:7:6: note: declared here
    7 | void g(){}
      |      ^
configure:6546: $? = 1
failed program was:
/* The following provokes a segfault in the compiler on powerpc-apple-darwin.
   Extracted from tests/mpn/t-iord_u.c.  Causes Apple's gcc 3.3 build 1640 and
   1666 to segfault with e.g., -O2 -mpowerpc64.  */

#if defined (__GNUC__) && ! defined (__cplusplus)
typedef unsigned long long t1;typedef t1*t2;
void g(){}
void h(){}
static __inline__ t1 e(t2 rp,t2 up,int n,t1 v0)
{t1 c,x,r;int i;if(v0){c=1;for(i=1;i<n;i++){x=up[i];r=x+1;rp[i]=r;}}return c;}
void f(){static const struct{t1 n;t1 src[9];t1 want[9];}d[]={{1,{0},{1}},};t1 got[9];int i;
for(i=0;i<1;i++){if(e(got,got,9,d[i].n)==0)h();g(i,d[i].src,d[i].n,got,d[i].want,9);if(d[i].n)h();}}
#else
int dummy;
#endif

int main () { return 0; }
configure:7030: result: no, long long reliability test 1
```

--- a/cross_compile_ffmpeg.sh
+++ b/cross_compile_ffmpeg.sh
@@ -345,6 +345,7 @@ install_cross_compiler() {
     if [[ ($compiler_flavors == "win32" || $compiler_flavors == "multi") && ! -f ../$win32_gcc ]]; then
       echo "Building win32 cross compiler..."
       download_gcc_build_script $zeranoe_script_name
+      download_gcc_build_script "gmp-gcc-15.patch"
       if [[ `uname` =~ "5.1" ]]; then # Avoid using secure API functions for compatibility with msvcrt.dll on Windows XP.
         sed -i "s/ --enable-secure-api//" $zeranoe_script_name
       fi
@@ -361,6 +362,7 @@ install_cross_compiler() {
     if [[ ($compiler_flavors == "win64" || $compiler_flavors == "multi") && ! -f ../$win64_gcc ]]; then
       echo "Building win64 x86_64 cross compiler..."
       download_gcc_build_script $zeranoe_script_name
+      download_gcc_build_script "gmp-gcc-15.patch"
       CFLAGS=-O2 CXXFLAGS=-O2 nice ./$zeranoe_script_name $zeranoe_script_options --build-type=win64 || exit 1
       if [[ ! -f ../$win64_gcc ]]; then
         echo "Failure building 64 bit gcc? Recommend nuke sandbox (rm -rf sandbox) and start over..."

--- a/patches/gmp-gcc-15.patch
+++ b/patches/gmp-gcc-15.patch
@@ -1,0 +1,21 @@
+
+# HG changeset patch
+# User Marc Glisse <marc.glisse@inria.fr>
+# Date 1738186682 -3600
+# Node ID 8e7bb4ae7a18b1405ea7f9cbcda450b7d920a901
+# Parent  e84c5c785bbe8ed8c3620194e50b65adfc2f5d83
+Complete function prototype in acinclude.m4 for C23 compatibility
+
+diff -r e84c5c785bbe -r 8e7bb4ae7a18 acinclude.m4
+--- a/acinclude.m4	Wed Dec 04 18:26:27 2024 +0100
++++ b/acinclude.m4	Wed Jan 29 22:38:02 2025 +0100
+@@ -609,7 +609,7 @@
+ 
+ #if defined (__GNUC__) && ! defined (__cplusplus)
+ typedef unsigned long long t1;typedef t1*t2;
+-void g(){}
++void g(int,t1 const*,t1,t2,t1 const*,int){}
+ void h(){}
+ static __inline__ t1 e(t2 rp,t2 up,int n,t1 v0)
+ {t1 c,x,r;int i;if(v0){c=1;for(i=1;i<n;i++){x=up[i];r=x+1;rp[i]=r;}}return c;}
+

--- a/patches/mingw-w64-build-r22.local
+++ b/patches/mingw-w64-build-r22.local
@@ -609,6 +609,17 @@ std_compile ()
 
   download_extract "$url"
 
+  # Patch for GMP on GCC 15
+  if [ "$pkg" = "gmp" ]; then
+    local gccver=$(gcc -dumpversion)
+    if [ "$gccver" -gt 14 ]; then
+      cd "$pkg-$ver"
+      patch -p1 < "../../../../gmp-gcc-15.patch"
+      autoreconf -vif
+      cd ".."
+    fi
+  fi
+
   if ! check_pkg_touch_success "$pkg" "$ver" 'configure' "$uname_m"; then
     clean_build
     build_progress "$pkg" 'Configuring'

